Add Accordions Block

jffng opened the pull request:
## What
Adds several experimental blocks to the block library to enable an [Accordion pattern](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/examples/accordion/). 

![Screenshot 2024-07-31 at 5 11 15 PM](https://github.com/user-attachments/assets/17396545-5b99-4c74-9cb5-2bc34c97da00)

**Video Demo**
https://cldup.com/transcoded/Oy55bzki_B.mov.mp4

## Why
See #21584 and #63501

## How 

- Implements four blocks: `Accordions` > `Accordion Item` > `Accordion Header` > `Accordion Panel`. This allows the user to style the trigger and content separately, while still maintaining the accessibility and semantic requirements of an accordion. I explored an approach with only two blocks — accordion group and item — but this made it difficult to provide the user with enough styling control over the trigger and content.
- Allows configuration of which items are open by default, if the drawers autoclose when another item is opened, and the icon.
- Uses the interactivity API + tag processor to handle interactivity in the view.
- I followed the HTML semantics and practices presented in these examples: [WCAG](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/examples/accordion/), [USWDS](https://designsystem.digital.gov/components/accordion/), and informed by design input from @jarekmorawski and @nyiriland. 

## Testing Instructions
- Visit Gutenberg > Experiments and check the option to "Enable experimental blocks"
- Verify you can implement an Accordion to a post, page or template. 

[Playground link to test the PR](https://playground.wordpress.net/#%7B%22landingPage%22:%22/wp-admin/%22,%22steps%22:%5B%7B%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22%7D,%7B%22step%22:%22mkdir%22,%22path%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22writeFile%22,%22path%22:%22/wordpress/pr/pr.zip%22,%22data%22:%7B%22resource%22:%22url%22,%22url%22:%22/plugin-proxy.php?org=WordPress&repo=gutenberg&workflow=Build%2520Gutenberg%2520Plugin%2520Zip&artifact=gutenberg-plugin&pr=64119%22,%22caption%22:%22Downloading%20Gutenberg%20PR%2064119%22%7D,%22progress%22:%7B%22weight%22:2,%22caption%22:%22Applying%20Gutenberg%20PR%2064119%22%7D%7D,%7B%22step%22:%22unzip%22,%22zipPath%22:%22/wordpress/pr/pr.zip%22,%22extractToPath%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22installPlugin%22,%22pluginZipFile%22:%7B%22resource%22:%22vfs%22,%22path%22:%22/wordpress/pr/gutenberg.zip%22%7D%7D%5D%7D) in browser. 

<details>

<summary>Test block markup</summary>

```
<!-- wp:accordions -->
<div class="wp-block-accordions"><!-- wp:accordion-item -->
<div class="wp-block-accordion-item"><!-- wp:accordion-header -->
<h3 class="wp-block-accordion-header accordion-item__heading"><button class="accordion-item__toggle"><span>Getting Started</span><span class="accordion-item__toggle-icon has-icon-plus" style="width:1.2em;height:1.2em"><span style="display:inline-block;width:1.2em;height:1.2em;line-height:1.2empx;text-align:center;font-size:NaNpx;font-weight:bold;color:currentColor">+</span></span></button></h3>
<!-- /wp:accordion-header -->

<!-- wp:accordion-panel -->
<div class="wp-block-accordion-panel"><div class="accordion-content__wrapper"><!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:accordion-panel --></div>
<!-- /wp:accordion-item -->

<!-- wp:accordion-item -->
<div class="wp-block-accordion-item"><!-- wp:accordion-header -->
<h3 class="wp-block-accordion-header accordion-item__heading"><button class="accordion-item__toggle"><span>Benefits of using accordion patterns</span><span class="accordion-item__toggle-icon has-icon-plus" style="width:1.2em;height:1.2em"><span style="display:inline-block;width:1.2em;height:1.2em;line-height:1.2empx;text-align:center;font-size:NaNpx;font-weight:bold;color:currentColor">+</span></span></button></h3>
<!-- /wp:accordion-header -->

<!-- wp:accordion-panel -->
<div class="wp-block-accordion-panel"><div class="accordion-content__wrapper"><!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:accordion-panel --></div>
<!-- /wp:accordion-item -->

<!-- wp:accordion-item -->
<div class="wp-block-accordion-item"><!-- wp:accordion-header -->
<h3 class="wp-block-accordion-header accordion-item__heading"><button class="accordion-item__toggle"><span>How to use the accordion block</span><span class="accordion-item__toggle-icon has-icon-plus" style="width:1.2em;height:1.2em"><span style="display:inline-block;width:1.2em;height:1.2em;line-height:1.2empx;text-align:center;font-size:NaNpx;font-weight:bold;color:currentColor">+</span></span></button></h3>
<!-- /wp:accordion-header -->

<!-- wp:accordion-panel -->
<div class="wp-block-accordion-panel"><div class="accordion-content__wrapper"><!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:accordion-panel --></div>
<!-- /wp:accordion-item -->

<!-- wp:accordion-item -->
<div class="wp-block-accordion-item"><!-- wp:accordion-header -->
<h3 class="wp-block-accordion-header accordion-item__heading"><button class="accordion-item__toggle"><span>Best practices of accordion patterns</span><span class="accordion-item__toggle-icon has-icon-plus" style="width:1.2em;height:1.2em"><span style="display:inline-block;width:1.2em;height:1.2em;line-height:1.2empx;text-align:center;font-size:NaNpx;font-weight:bold;color:currentColor">+</span></span></button></h3>
<!-- /wp:accordion-header -->

<!-- wp:accordion-panel -->
<div class="wp-block-accordion-panel"><div class="accordion-content__wrapper"><!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:accordion-panel --></div>
<!-- /wp:accordion-item -->

<!-- wp:accordion-item -->
<div class="wp-block-accordion-item"><!-- wp:accordion-header -->
<h3 class="wp-block-accordion-header accordion-item__heading"><button class="accordion-item__toggle"><span>Common questions and mistakes</span><span class="accordion-item__toggle-icon has-icon-plus" style="width:1.2em;height:1.2em"><span style="display:inline-block;width:1.2em;height:1.2em;line-height:1.2empx;text-align:center;font-size:NaNpx;font-weight:bold;color:currentColor">+</span></span></button></h3>
<!-- /wp:accordion-header -->

<!-- wp:accordion-panel {"isSelected":true} -->
<div class="wp-block-accordion-panel"><div class="accordion-content__wrapper"><!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:accordion-panel --></div>
<!-- /wp:accordion-item --></div>
<!-- /wp:accordions -->
```

</details>

jffng committed:
Add initial accordion implementation.

jffng committed:
Refactor blocks for core loading.

jffng committed:
Simplify state getter by storing id in the context.

luisherranz commented on packages/block-library/src/accordion-group/index.php:
```diff
@@ -0,0 +1,59 @@
+	$p         = new WP_HTML_Tag_Processor( $content );
+	$autoclose = $attributes['autoclose'];
+
+	while ( $p->next_tag() ) {
+		if ( $p->has_class( 'wp-block-accordion-group' ) ) {
+			$p->set_attribute( 'data-wp-interactive', 'core/accordion' );
+			$p->set_attribute( 'data-wp-context', '{"isOpen":[],"autoclose":"' . $autoclose . '"}' );
+		}
+	}
```
If I'm not mistaken, you should be able to do this.

```suggestion
	if ( $p->next_tag( array( 'class_name' => 'wp-block-accordion-group' ) ) ) {
    $p->set_attribute( 'data-wp-interactive', 'core/accordion' );
    $p->set_attribute( 'data-wp-context', '{"isOpen":[],"autoclose":"' . $autoclose . '"}' );
	}
```

luisherranz commented on packages/block-library/src/accordion-group/index.php:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	$p         = new WP_HTML_Tag_Processor( $content );
+	$autoclose = $attributes['autoclose'];
```
This should probably be escaped or converted to a boolean to prevent people from injecting nasty things into the JSON.

```suggestion
	$autoclose = (bool) $attributes['autoclose'];
```

luisherranz commented on packages/block-library/src/accordion-group/view.js:
```diff
@@ -0,0 +1,34 @@
+			} else if ( itemIsOpen ) {
+				context.isOpen = isOpen.filter( ( item ) => item !== id );
+			} else {
+				context.isOpen = [ ...isOpen, id ];
```
You don't need to use immutability with the Interactivity API. In fact, it's usually better not to use it for performance reasons.

```suggestion
				context.isOpen.push( id );
```

luisherranz commented on packages/block-library/src/accordion-item/index.php:
```diff
@@ -0,0 +1,69 @@
+	while ( $p->next_tag() ) {
+		if ( $p->has_class( 'wp-block-accordion-item' ) ) {
+            $p->set_attribute( 'data-wp-context', '{ "id": "' . $unique_id . '" }' );
+			$p->set_attribute( 'data-wp-class--is-open', 'state.isOpen' );
+			if ( $attributes['openByDefault'] ) {
+				$p->set_attribute( 'data-wp-init', 'callbacks.open' );
+			}
```
As you are using JavaScript to open the item, the initial HTML sent to the browser will have the item closed and will only open after the JavaScript loads, causing a layout shift. 

It would be better if the state on the server could already contain the correct information about which item is opened so that the HTML the browser receives is the correct HTML.

There are several ways to share information between blocks. One option could be something similar to what [we've done in the image block](https://github.com/WordPress/gutenberg/pull/63348), where we put the info in the global state and only an `id` in the local context. But it's not the only way, so feel free to choose the best one for you.

_(By the way, I am preparing a new guide on server-side rendering. Once I open the pull request, I will put a link here.)_

jffng replied:
> It would be better if the state on the server could already contain the correct information about which item is opened so that the HTML the browser receives is the correct HTML.

I tried to initialize the state on the server: https://github.com/WordPress/gutenberg/pull/64119/commits/4c27428623c99430d65d4cf7ade64a233a9dca14

There are two issues now:

1. I'm still getting layout shift, I don't know why this is the case since I thought the state is being initialized with the right info on the server.
2. It breaks if there is more than one accordion group on the page and `autoclose` (only allows one item per group to be open at a time) is enabled, since they are sharing the same store and I'm using global state to track which ids are open.

Do you have any suggestions? Thank you for the review so far.

luisherranz replied:
> I'm still getting layout shift, I don't know why this is the case since I thought the state is being initialized with the right info on the server.

The derived state was missing on the server. 

> It breaks if there is more than one accordion group on the page and autoclose (only allows one item per group to be open at a time) is enabled, since they are sharing the same store and I'm using global state to track which ids are open.

To fix the autoclose issue, I refactored the code to move the array with the IDs of open items to the local context.

I've made a video to explain the two changes.

<div>
    <a href="https://www.loom.com/share/8b44e78546b7415bba02de825f59c045">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/8b44e78546b7415bba02de825f59c045-ccb230eecab2ed0e-full-play.gif">
    </a>
  </div>

https://www.loom.com/share/8b44e78546b7415bba02de825f59c045

jffng replied:
Thanks for the updates and video! Super helpful, and it's very useful to see how the local context is initialized on the server using a closure.

luisherranz commented on packages/block-library/src/accordion-item/index.php:
```diff
@@ -0,0 +1,69 @@
+		}
+	}
+
+	$content = $p->get_updated_html();
+	$p       = new WP_HTML_Tag_Processor( $content );
+
+	while ( $p->next_tag() ) {
+		if ( $p->has_class( 'accordion-item__toggle' ) ) {
```
I don't think you need to rewind unless the order of appearance of the tags is unknown.

Ultimately, if it doesn't find a tag, there's no need to add the following ones either, since the block will be broken anyway, won't it?

I would just nest the checks in the correct order, something like this:

```php
if ( $p->next_tag( array( 'class_name' => 'wp-block-accordion-group' ) ) ) {
  // Add directives..
  if ( $p->next_tag( array( 'class_name' => 'accordion-item__toggle' ) ) ) {
    // Add directives..
    if ( $p->next_tag( array( 'class_name' => 'wp-block-accordion-content' ) ) ) {
      // Add directives..

      // If everything went well, overwrite the content.
      $content = $p->get_updated_html();
    }
  }
}
```

If the order is unknown, then you can use a bookmark to rewind:

```php
$p->set_bookmark('start');

if ( $p->next_tag( array( 'class_name' => 'wp-block-accordion-group' ) ) ) {
  // Add directives..

  // Rewind
  $p->seek('start');
  if ( $p->next_tag( array( 'class_name' => 'accordion-item__toggle' ) ) ) {
    // Add directives..

    // Rewind
    $p->seek('start');
    if ( $p->next_tag( array( 'class_name' => 'wp-block-accordion-content' ) ) ) {
      // Add directives..

      // If everything went well, overwrite the content.
      $content = $p->get_updated_html();
    }
  }
}
```

luisherranz replied:
Solved in https://github.com/WordPress/gutenberg/pull/64119/commits/4c27428623c99430d65d4cf7ade64a233a9dca14.

luisherranz reviewed (commented):
This is exciting. I'm happy to see more interactive blocks in Core 🙂 

I have left some comments here, but overall it looks great.

jffng committed:
Clean up tag processing and try to initialize state on the server.

luisherranz committed:
Don't enqueue the frontend files if there's no content

luisherranz committed:
Stringify the autoclose value

luisherranz committed:
Refactor to populate `context.isOpen` on the server

luisherranz commented on packages/block-library/src/accordion-group/index.php:
```diff
@@ -0,0 +1,73 @@
+	wp_enqueue_script_module( '@wordpress/block-library/accordion-group' );
+
+	$p         = new WP_HTML_Tag_Processor( $content );
+	$autoclose = $attributes['autoclose'] ? 'true' : 'false';
```
As we're not using `wp_json_encode`, we need to perform the boolean stringify ourselves.

Fixed in https://github.com/WordPress/gutenberg/pull/64119/commits/90a7c66db1174b44f49d35e41f314c866a200d5a.

luisherranz commented on packages/block-library/src/accordion-item/index.php:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	$p         = new WP_HTML_Tag_Processor( $content );
+	$unique_id = wp_unique_id( 'accordion-item-' );
```
Let's leave it with `wp_unique_id` for the moment 👍

luisherranz commented on packages/block-library/src/accordion-item/index.php:
```diff
@@ -0,0 +1,73 @@
+	// Adds the id to the current accordion group list.
+	if ( $attributes['openByDefault'] ) {
+		// phpcs:ignore Gutenberg.CodeAnalysis.ForbiddenFunctionsAndClasses.ForbiddenFunctionCall
+		gutenberg_block_core_accordion_group_item_ids( $unique_id );
```
It is necessary to review how to share functions between different blocks.

luisherranz reviewed (changes requested):
Code-wise seems good to me. It'll probably be ready to go once you've reviewed how to share functions between different Gutenberg blocks 🙂

EDIT: I've labeled the pull request with "Needs Accessibility Review" to ensure this block also undergoes an accessibility review.

jffng committed:
Add comments to the server side rendering functions.

jffng commented:
I plan to follow up with some design and UX improvements, and will mark the PR ready for review after that.

jffng committed:
Improve the block editing UX

- Label the None icon setting
- Add help text to block settings
- Remove button focus styles
- Add border support to the group.

Nyiriland commented:
@jffng Picking up convo [from over here](https://github.com/a8cteam51/special-projects-blocks-monorepo/pull/16#issuecomment-2272221815).

> do you still think h4 or h5 is best for the default, or can we keep it h3?

I polled some other designers who regularly build sites, and the vote was to keep it an h3 (although both work IMO). I have a feeling h5 will most likely be too small in most themes.

> Also something to consider is when this setting is enabled and multiple accordion items are open by default, clicking any open section actually closes all of them.

Yeah, although this is a bit weird, it makes functional sense, and should be included in the wording. What about this for the `Accordion Group` language: 

**Toggle:** "Auto-close"
**Helper text:** "Automatically closes accordions when a new one is opened."

jffng committed:
Reword the auto-close toggle help text.

jffng committed:
Add padding to the content and trigger by default.

ndiego commented:
Just noting that it would be good to make the Heading level attribute extensible as we recently did with other Core blocks that use the `HeadingLevelDropdown` component. 

Here's the original PR for reference: https://github.com/WordPress/gutenberg/pull/63535

hanneslsm commented:
> > do you still think h4 or h5 is best for the default, or can we keep it h3?

> I polled some other designers who regularly build sites, and the vote was to keep it an h3 (although both work IMO). I have a feeling h5 will most likely be too small in most themes.

I agree that h3 can be default because usually the sections that have an accordion have a h2. 
However, I feel like it needs to be noted (saying this as a designer): Styling shouldn't be the reason for this decision but semantics. h3 could also too big and besides font-size there are many ways to create hierarchy of the headlines (e.g. text-transform, color, borders). But, as said, considering semantics, h3 is a good choice.

mikachan commented on packages/block-library/src/accordion-item/index.php:
```diff
@@ -0,0 +1,76 @@
+			$p->set_attribute( 'data-wp-bind--aria-expanded', 'state.isOpen' );
+
+			if ( $p->next_tag( array( 'class_name' => 'wp-block-accordion-content' ) ) ) {
+				$p->set_attribute( 'aria-labelledby', $unique_id );
```
I believe `aria-labelledby` should reference an `id` attribute, so we may need to set an `id` alongside the `aria-controls` on line 47, e.g.

```
$p->set_attribute( 'id', $unique_id );
```

Also, I think `aria-controls` should reference the `id` of the section/contents that it is controlling, i.e. the `id` attribute of the `wp-block-accordion-content`. We'd need a different ID to the one used by `aria-labelledby`, so perhaps the label ID could be something like `accordion-trigger-item-x`, and the content ID could be `accordion-content-item-x`.

I'm basing this off [the example here](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/examples/accordion/).

jffng replied:
Good catch, fixed here: [88428a6](https://github.com/WordPress/gutenberg/pull/64119/commits/88428a676911e577cdf5bc420ddd1bbd52f7bbf2)

mikachan reviewed (commented):
This is looking amazing, thank you for working on it ✨ 

When testing it out, I noticed that the accordions are difficult to close in the Editor when clicking on the close button, but maybe this is on purpose, as you also need to be able to click to edit the title:

https://github.com/user-attachments/assets/7886a777-502e-4067-8c71-8e3b36f6cd66

They close easily if I click anywhere outside the accordion content.

Otherwise, this is testing on the front-end and in the Editor really well for me. 

I've also tested using a screen reader, and I believe all the notices that are read out sound correct. I am also able to tab through the accordions in the Editor to edit their titles and contents. However, it would be great to see more specific accessibility feedback before this lands. I did notice a small potential change to the way `aria-labelledby` and `aria-controls` are used, I think both may need their own unique `id` references.

luisherranz commented on packages/block-library/src/accordion-group/index.php:
```diff
@@ -0,0 +1,80 @@
+ * @param string $id The id of the accordion item.
+ * @return array|void Returns the ids for the  or nothing.
+ */
+function block_core_accordion_group_item_ids( $id = null ) {
```
@gziolo Could you help us with the name of this function? How is it supposed to be named so that it can be used from the other block? What pattern could be used here?

luisherranz commented on packages/block-library/src/accordion-item/index.php:
```diff
@@ -0,0 +1,76 @@
+	// add its id to the accordion group's initial context.isOpen list.
+	if ( $attributes['openByDefault'] ) {
+		// phpcs:ignore Gutenberg.CodeAnalysis.ForbiddenFunctionsAndClasses.ForbiddenFunctionCall
+		gutenberg_block_core_accordion_group_item_ids( $unique_id );
```
We need to use it in this other block, or find an alternative pattern.

jffng replied:
I tried an alternative in https://github.com/WordPress/gutenberg/pull/64119/commits/3e7cae6a488daf4eba61b06653696958d11e1eb5 which alters how the value of `state.isOpen` is initialized on the server based on the block attribute (`openByDefault`), and uses a callback to ensure the context matches, avoiding the shared function. What do you think?

luisherranz replied:
It's not so clean in terms of state/context, but if that works, it's fine 🙂

By the way, I think it was me who suggested using `(bool)`, but I think it's not right; it's better to use the strings `true` or `false` without quotes, so that they are a boolean property in the generated JSON. Something like this:

```php
$open_by_default = $attributes['openByDefault'] ? 'true' : 'false';
$p->set_attribute( 'data-wp-context', '{ "id": "' . $unique_id . '", "openByDefault": ' . $open_by_default . ' }' );
```

jffng replied:
Addressed in [9ab8033](https://github.com/WordPress/gutenberg/pull/64119/commits/9ab80339c66dc635ffede2785578ff7e5b93e1c8).

luisherranz reviewed (dismissed):
Just as a reminder, we need to fix this before merging, as we can't include a function that starts with `gutenberg_`, since that doesn't work in WordPress Core.

jffng committed:
Fix aria-labelledby and aria-controls references.

jffng committed:
Allow HeadingLevelOptions to be extended.

jffng commented:
> Just noting that it would be good to make the Heading level attribute extensible as we recently did with other Core blocks that use the HeadingLevelDropdown component.

Done in [aa687ca](https://github.com/WordPress/gutenberg/pull/64119/commits/aa687caf375c874459966b218b8c91dab99fcd0d)

> When testing it out, I noticed that the accordions are difficult to close in the Editor when clicking on the close button, but maybe this is on purpose, as you also need to be able to click to edit the title:

Yes that's intentional right now — if an `Accordion Item` or any its children are selected, the content is visible.

> Just as a reminder, we need to fix this before merging, as we can't include a function that starts with gutenberg_, since that doesn't work in WordPress Core.

Yes, I'm thinking about an alternative approach that doesn't rely on a shared function / static variable between the two blocks. Right now it's not the easiest to understand. 

Once that's done I can prepare a Core backport PR.

joedolson commented on packages/block-library/src/accordion-content/edit.js:
```diff
@@ -0,0 +1,61 @@
+				...colorProps.style,
+				...shadowProps.style,
+			} }
+			aria-hidden={ ! isSelected && ! openByDefault }
```
Using `aria-hidden="true"` on concealed content will run into accessibility problems if the contained content has focusable content. I assume this is used to avoid `display: none`, since that causes problems with animating the state change, but it will also introduce accessibility problems. 

Might work if you find all focusable elements inside the content container and disable their keyboard navigability when aria-hidden is applied.

jffng replied:
Yes that's exactly right. I will try your suggestion of disabling keyboard navigability of focusable elements when aria-hidden is applied!

jffng replied:
> find all focusable elements inside the content container and disable their keyboard navigability when aria-hidden is applied.

@joedolson I understand this in the context of the front-end experience, but I'm wondering how it should be handled in the editor. The aria-hidden property is removed if the accordion item or any of its children are selected, which makes it so the contained content is always effectively keyboard navigable: 

https://github.com/user-attachments/assets/c88ac4b8-88e0-444d-9659-d6dd5a3cf7ec

jffng replied:
I addressed this in the front-end view: https://github.com/WordPress/gutenberg/pull/64119/commits/b522ddad5840bde81ae40be2a754d53a02eab518

@luisherranz I first tried to implement this on the server by by setting `data-wp-bind--tabindex` to the derived state, but because of how many blocks (like the file or navigation block) are rendered, the tabindex was being overridden, thus still ended up in the tab order. Curious if you have any thoughts or if what I've done above is a decent solution.

joedolson replied:
If there's no keyboard path to reach the hidden content that doesn't cause it to become visible, then it's probably not necessary to also hide it in the editor. That should probably get checked over with a few different screen readers, however; I think that a screen reader user navigating with the virtual cursor would still be able to get to those items.

jffng replied:
So the thought is that we can remove the toggling of `aria-hidden` in the editor, since there is no keyboard path to reach the hidden content that doesn't cause it to become visible?

luisherranz replied:
> the tabindex was being overridden, thus still ended up in the tab order

I did a quick test, and it seems to work well. What steps do I need to take to reproduce the problem you describe?

Although I am not an accessibility expert,  maybe a simpler option would be to use the `inert` attribute with the `data-wp-bind` directive on the accordion item.

`data-wp-bind--inert="!state.isOpen"`

joedolson replied:
Using `inert` is a good call; I'm not in the habit of suggesting it, since it's so newly available.

jffng replied:
Nice suggestion, using `insert` works as expected — child elements are not keyboard navigable when the accordion item is collapsed. Applied this change in [aebf612](https://github.com/WordPress/gutenberg/pull/64119/commits/aebf6125d9e9ae84239141bf5dca2c694c256364).

jffng committed:
Only allow bold, italic, image, and strikethrough for the Trigger.

jffng committed:
Disable keyboard navigation for focusable elements when aria-hidden is true.

jffng committed:
Fix __next40pxDefaultSize linting error.

jffng committed:
Refactor how state and context are initialized to avoid the need for a shared function between blocks.

jffng commented:
There are a few pieces of outstanding feedback: 

- Do we need to apply the `aria-hidden` attribute in the editor? https://github.com/WordPress/gutenberg/pull/64119#discussion_r1720048188
- Is [this](https://github.com/WordPress/gutenberg/pull/64119#discussion_r1720136283) approach to initializing the state okay? https://github.com/WordPress/gutenberg/pull/64119#pullrequestreview-2237981510
- If this block is behind a Gutenberg experiments flag, should there be a core backport PR? I think it could make sense to prepare the core backport PR when the block is moved out from the experiments flag. 

Heads up that I'm AFK until August 28. I plan to wrap this up when I return to the keyboard, unless someone else wants to help bring it in before then :)

jffng committed:
Use strings for openByDefault so boolean is in the generated JSON.

jffng committed:
Use inert instead of aria-hidden.

jffng commented:
I've addressed the outstanding feedback, and added the `No Core Sync Required` tag since this is behind an experimental feature flag. It is ready for another review.

gziolo commented:
A quick high-level feedback from looking at [Accordion Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/) as presented at ARIA Authoring Practices:

> Terms for understanding accordions include:
>
> **Accordion Header:**
> Label for or thumbnail representing a section of content that also serves as a control for showing, and in some implementations, hiding the section of content.
> **Accordion Panel:**
> Section of content associated with an accordion header.

I would be in favor of naming the block accordingly instead of Trigger and Content as these names presented in the UI might be misleading. In particular, there is already Post Content block that is presented in the UI as Content, so there would be ambiguity.

> Implements four blocks: Accordion Group > Accordion Item > Accordion Trigger > Accordion Content. This allows the user to style the trigger and content separately, while still maintaining the accessibility and semantic requirements of an accordion. I explored an approach with only two blocks — accordion group and item — but this made it difficult to provide the user with enough styling control over the trigger and content.

In https://github.com/WordPress/gutenberg/pull/63689, @creativecoder works on the Tabs block. Do you think the same rationale should drive the final implementation of that block that initially was coded using only two blocks Tabs and Tab, which would be similar to having Accordion Group and Accordion. By the way, the Details block, which behaves like a pair of accordion header and panel combined together, is a single block with inner blocks provided by the user. Concluding, it would be good to establish a similar decision making process for all core blocks so the final user experience is coherent for users.

carolinan commented:
I support using the multiple blocks, the details block also started out with multiple blocks and more style options, but was reduced on request by reviewers. Which is one of the reason why we need a more flexible accordion block.

creativecoder commented:
> Do you think the same rationale should drive the final implementation of that block that initially was coded using only two blocks Tabs and Tab, which would be similar to having Accordion Group and Accordion.

Note that I anticipate needing to break out the tab block into tab-label and tab-panel. Because the style supports system is set up to add controls for the block as a whole, I think we'll need to separate these so they can have independent style controls (background, border, margin/padding, etc)

jffng committed:
Rename Trigger and Content to Header and Panel.

jffng committed:
Rename interactivity to script modules.

jffng committed:
Remove unused watch directive for setting the tabIndex.

MaggieCabrera commented:
This is looking good, but I think we should move the icon selection one level up, else if we want a non default icon, we need to change it manually for every trigger instead

LukaszJaro commented:
The playground PR is throwing error so just wondering, will it be possible to add custom ID to the trigger?

mikachan committed:
Merge branch 'trunk' into try/accordion-block

mikachan committed:
Move icon controls to top level

mikachan committed:
Fix linting

mikachan committed:
Remove old GB experiments

MaggieCabrera commented:
I'm testing quickly here, I have a question: how often are we expecting users to style each accordion group differently from each other? This is behaving just like the list item block, so I don't think this should be a blocker but we might want to ask ourselves if we should have an option to sync styles in between items, to keep them consistent and avoid having to replicate styles for all of them (since these items are likely going to be styled more often than the list item would, having borders and such). 

Have we had a design pass at the default CSS for this? I think the accordion panel should have the same spacing that the title does, but that's a personal opinion:

<img width="793" height="426" alt="Screenshot 2025-08-07 at 10 43 04" src="https://github.com/user-attachments/assets/4f62d948-3bbc-4dec-a1b2-3238cd89ad60" />

mikachan commented:
> how often are we expecting users to style each accordion group differently from each other? 

I'm not sure; I imagine this wouldn't be the most common use case, but it's still a possibility. Being able to sync styles as a follow-up sounds like a great idea.

> Have we had a design pass at the default CSS for this?

I think there was design input from @jarekmorawski and @Nyiriland when this PR was first opened, but I imagine there hasn't been any recent design reviews. Do you mean the accordion panel should have less spacing, to match the spacing between Title and Content? When the panels are all closed, they do look a little too far apart, but just checking which spacing you mean.

MaggieCabrera commented:
> I think there was design input from @jarekmorawski and @Nyiriland when this PR was first opened, but I imagine there hasn't been any recent design reviews. Do you mean the accordion panel should have less spacing, to match the spacing between Title and Content? When the panels are all closed, they do look a little too far apart, but just checking which spacing you mean.

I mean that it's inconsistent between the two. In a blank theme the title seems to have too much space and the content none at all. I'd say they should have roughly the same, even if it's none at all.

/cc @jasmussen in case he wants to chime in

jasmussen commented:
> I mean that it's inconsistent between the two. In a blank theme the title seems to have too much space and the content none at all. I'd say they should have roughly the same, even if it's none at all.

Is there a good way to test, or do you happen to have some small videos handy? I imagine there may be some inherited styles in some of these states we might want to be sure to control.

Many properties I'd love surfaced as style options, border, color, spacing, typography, etc, but even then we need some baseline defaults.

MaggieCabrera commented:
> > The panel content is using the default `p` spacing inherited from the browser, and the title is using custom padding of `var(--wp--preset--spacing--20, 1em) 0;`.
> 
> Interesting, for me the title in empty theme doesn't have margin or padding at all:

Nevermind that! there's padding on the button inside the h3

jasmussen commented:
Took this for a spin. Incredibly impressive work, kudos. Here's what I see:

![accordion test](https://github.com/user-attachments/assets/5a0ac748-c3d0-4d49-98a1-07db75fab130)

Honestly this is working pretty well and much of this can arguably ship as is. The styling options are even fairly reasonable and inherit well:

<img width="841" height="506" alt="image" src="https://github.com/user-attachments/assets/f299f7dc-c1f9-4326-9501-1354a289b8f8" />

There's one aspect I'm waffling on, however, which is this one: 
<img width="272" height="96" alt="image" src="https://github.com/user-attachments/assets/f3e02c20-e321-47fa-a2f8-9294beb09f66" />

I think we should extract the custom icon feature to a separate branch, so that we can ship a version of this PR without those icons. Probably we can add a unicode glyph instead, some alternative. I realize this may be disappointing feedback, as it's working well as is, so let me elaborate on why.

* The icons you've included, although good, are admin-UI icons, not specifically designed for the front-end. 
* Even if they are good, it's just 5 icons, and there's no customization beyond that. 
* If we ship with those icons in a version of WordPress, we have to support them forever, even if we find a better alternative in the future.
* The icon tech is not entirely up to date compared to modern standards, where icon sets like Lucide provide customizable stroke-width and optical footprint, vastly expanding the range of options for a theme.
* We've even had feedback for the icon customization aspect of the navigation menu as well as the Details block, that simply providing a unicode glyph instead of an icon would provide options for customization beyond what a preset collection of icons would.
* It's on my mind that [this issue is picking up steam](https://github.com/WordPress/gutenberg/issues/71026#issue-3284283822), and it specifically details an "Icon Block", which is meant to solve all the items I mentioned here.

In other words I'm not saying we don't want icons. What I'm saying is that we want a vastly expanded range of icons with more customization than we are able to provide with the WordPress set, and so instead of shipping a collection of 5, even if they are good, we should find a way to ship without icons so that we can update with this customizability in the future.

What do you think?

jasmussen commented:
> What do you think would work best as the alternative? 

The icon rotating to become a close button is a good experience. Could you use the [plus](https://symbl.cc/en/002B/) (`+`) instead?

> And do you think we should we keep the "None" option?

In a way, yes! Other quirks of the navigation block aside, "Show arrow" works reasonably well:

<img width="161" height="99" alt="image" src="https://github.com/user-attachments/assets/af937126-09c1-4f05-b451-086040079f95" />

In this case, it could be a "Show icon" toggle. The perspective here could then be, in the future when we return to this, the controls to select icon and glyph could be contextual to this toggle, hidden if toggled off. What do you think?

mikachan committed:
Remove icon selection

mikachan committed:
Select first Accordion Header by default

mikachan committed:
Fix icon logic in save.js

MaggieCabrera commented:
RTL works out of the box:

<img width="1324" height="745" alt="Screenshot 2025-08-08 at 12 05 51" src="https://github.com/user-attachments/assets/96ac073e-5fb1-4813-84b1-abbe06d61eb8" />

mikachan committed:
Remove layout controls from header and panel

mikachan committed:
Fix RTL icon position

jasmussen commented:
As far as layout tools, _if it's simpler to start with less_, that could be a path forward. Presuming you can insert a Group block you can always use layout tools from that, yes?

mikachan committed:
Remove layout from group and item

mikachan committed:
Add layout to accordion group

mikachan commented:
> As someone who builds out sites and pages regularly, I would definitely recommend keeping layout width controls on the parent accordion group, as https://github.com/WordPress/gutenberg/pull/64119#issuecomment-3167625009.

Gotcha, that makes sense, thanks for the feedback! I've added the layout control back to the Accordion Group in https://github.com/WordPress/gutenberg/pull/64119/commits/d4bb0e4c6e910ae6ed8b98c67d8c284aecce8c95.

mikachan committed:
Rename Accordion Group to Accordions

mikachan commented:
Thanks, @jasmussen, I agree! I've renamed the block to Accordions in https://github.com/WordPress/gutenberg/pull/64119/commits/f713c57d7bf9455f379ed20b6e16f831bd4b7f3a.

jasmussen commented:
From my POV this one is ready! Just needs a good code gutcheck.

scruffian commented on packages/block-library/src/accordion-item/edit.js:
```diff
@@ -0,0 +1,110 @@
+		[ clientId ]
+	);
+
+	const getBlockOrder = useSelect(
```
I think it would be better to select `isSelected` and `getBlockOrder` in the same useSelect (see https://developer.wordpress.org/news/2024/03/how-to-work-effectively-with-the-useselect-hook/#do-all-selections-from-the-same-store-in-one-callback)

mikachan replied:
Thanks! Done in https://github.com/WordPress/gutenberg/pull/64119/commits/fb546f792bae417a1185c846789789308dede82d.

scruffian commented on packages/block-library/src/accordions/index.js:
```diff
@@ -0,0 +1,49 @@
+import metadata from './block.json';
+import initBlock from '../utils/init-block';
+
+const icon = (
```
I wonder if these icons should be added to the library?

mikachan replied:
That makes sense, I've moved them to the icon library in https://github.com/WordPress/gutenberg/pull/64119/commits/0be4b657c93974f01023a039cf576a7b6e5cf446.

scruffian reviewed (changes requested):
I tested this with Global Styles. When I set a color for the heading, the CSS for the block overrides it:

<img width="686" height="633" alt="Screenshot 2025-08-13 at 14 01 48" src="https://github.com/user-attachments/assets/02d1b59e-91b9-472c-858f-465ab237967a" />

scruffian commented on packages/block-library/src/accordions/style.scss:
```diff
@@ -0,0 +1,92 @@
+	grid-template-rows: max-content 1fr;
+}
+
+.wp-block-accordion-item .accordion-item__heading {
```
It looks like these lines are to prevent heading element styles from bleeding into this block. The problem is that they prevent the global styles for the block from working. 

FWIW I think it's ok for this block to inherit heading element styles, since it uses an element. If designers want to only target heading blocks they can do so.

mikachan replied:
Good spot, thanks for testing that. I think we can safely remove the color and padding properties, but without the `margin: 0`, the headings have quite a lot of extra space from the browser defaults:

<img width="1380" height="700" alt="image" src="https://github.com/user-attachments/assets/58162b27-0291-4973-9d91-6495a5441d54" />

I've tried replacing this with `margin-block-start: 0;` and `margin-block-end: 0;`, maybe this will work better:

| Before | After |
| ----- | ----- |
| <img width="702" height="248" alt="image" src="https://github.com/user-attachments/assets/ccc7f115-4c7d-4c0f-a713-e9b3ca9b1e78" /> | <img width="702" height="248" alt="image" src="https://github.com/user-attachments/assets/eb23a913-dd3f-4281-96ef-0db5dff1593c" /> |

scruffian replied:
These rules still override whatever is set in Global Styles. I see three options:
1. Don't allow users to change margin on these blocks.
2. Make the CSS less specific.
3. Set default values explicitly.

mikachan replied:
I think it'd best to keep the margin controls on these blocks, so I've tried making the CSS less specific in https://github.com/WordPress/gutenberg/pull/64119/commits/43dc276227066b8e9b59be99873d9dedca8f9b5d. This means any margins set in Global Styles should be applied correctly.

scruffian replied:
It works!

mikachan committed:
Use single useSelect

mikachan committed:
Move accordion icons to icon library

mikachan committed:
Remove color and padding from heading styles

mikachan committed:
Make heading CSS less specific

scruffian reviewed (approved):
This is looking good to me now. Let's bring it in!

jasmussen commented:
> Have we considered using a Heading block inside the Accordion Header? By doing so, we won't have to worry about styling the heading anymore. It also paves the way for extenders to customize the header content.

The main challenge is that it might make it easy to create incorrect heading level hierarchies in the DOM. E.g. usually your post title is an H2, so we can default H3 in the accordion, which should make it match in _most cases_. However then you insert an accordion on your homepage, where perhaps the site title is an H1, then you'll have a missing level in-between.

This is usually the reason to not use heading blocks in compound blocks, because you have to choose a heading level and often times that can cause conflicts. Whereas using just a paragraph block will enable any styling, even to match headings or intentionally diverge, and it will never cause such hierarchy conflicts.

MaggieCabrera commented:
> That's valid enough. Another thing on the radar is a "linting tool", a la VS Code, but for the block editor. It's not a new idea and likely exists in an issue in the archive somewhere.
> 
> This would provide an interface for flagging warnings, errors, and notice level bits. Incorrect headings should be flagged as errors, and could well show up in a pre-publish dialog, as well as the document outline. I.e. if an error is shown, we’d show an unread dot on top of the List view button (which serendipitously would be similar to [the unread dot here](https://github.com/WordPress/gutenberg/issues/50756#issuecomment-3187347960)), alerting you to the issue.

I was thinking about this too. It would need the context of the whole template, if editing on the post editor, but it would be really helpful!

MaggieCabrera commented:
> Yes, this would be a global linting system for anything, not just pieces here or there. Choosing the right heading levels is not a challenge unique to just one block. The system could also flag in the three levels many other things that would be good to know about, whether failing contrasts, missing or even dead links, blocks in a setup state, hidden blocks, etc etc. The system should also be extensible, so potentially SEO plugins can use an official API to provide suggestions.
> 
> But to be clear, I would not think of this system as a blocker for the accordion to land. This idea has been around for years at this point, and although its potential value keeps going up, it hasn't yet been more important than many of the other shortcomings we mean to address, including this block itself!

yes yes, I didn't mean to derail this PR :) I would love to see this happening though, I hope we can prioritize it soon enough

mikachan commented:
> There is one use case that comes from WooCommerce: previously, with the classic theme, we used a tab for product details and displayed the number of reviews in the tab title of the review tab. Recently, we introduced a new accordion layout (which is a clone of this block), but we can't show the number of reviews in the accordion header because it's a static string set in the editor. If the Accordion Header block supports inner blocks, we can solve that issue nicely with a custom block. I will have the same need if I wrap the Comments block inside an Accordion.

I think the biggest issue with allowing inner blocks inside the Accordion Header is that it could disrupt the semantic markup of this block. However, I'll have a think about how we could address this. Maybe we could allow appending string content to the existing header string.

Thanks for all the feedback! I'll get this merged, and then we can start addressing the follow-ups:

- Add an "Add" button to the toolbar to easily add another Accordion Item
- Think about how to append content to the Accordion Header content

portrayaloflife commented:
Thank you everyone!

mikachan commented:
@sntran It's usually best to [open a new issue](https://github.com/WordPress/gutenberg/issues/new?template=Feature_request.md) for feature requests like this. This sounds like an interesting idea, especially for FAQ patterns.

ellatrix commented:
Why aren't we using the HTML native details element for this? Which stylistic limitations can't be overcome?
